macro/src/wasm_loader.rs

```diff
@@ -57,3 +46,115 @@ fn maybe_decompress(file_contents: Vec<u8>) -> WasmMetadataResult<Vec<u8>> {
         .map_err(|e| CodegenError::Wasm(e.to_string()))
         .map(Cow::into_owned)
 }
+
+struct Executor {
+    runtime_blob: RuntimeBlob,
+    executor: WasmExecutor<sp_io::SubstrateHostFunctions>,
+    externalities: sp_state_machine::BasicExternalities,
+}
+
+impl Executor {
+    fn new(wasm_file: &[u8]) -> WasmMetadataResult<Self> {
+        let externalities: sp_state_machine::BasicExternalities = Default::default();
+
+        let executor: WasmExecutor<sp_io::SubstrateHostFunctions> = WasmExecutor::builder()
+            .with_execution_method(WasmExecutionMethod::default())
+            .with_offchain_heap_alloc_strategy(sc_executor::HeapAllocStrategy::Dynamic {
+                maximum_pages: Some(64),
+            })
+            .with_max_runtime_instances(1)
+            .with_runtime_cache_size(1)
+            .build();
+
+        let runtime_blob =
+            RuntimeBlob::new(wasm_file).map_err(|e| CodegenError::Wasm(e.to_string()))?;
+
+        Ok(Self {
+            runtime_blob,
+            executor,
+            externalities,
+        })
+    }
+
+    fn versions(&mut self) -> WasmMetadataResult<Vec<u32>> {
+        let version = self
+            .executor
+            .uncached_call(
+                self.runtime_blob.clone(),
+                &mut self.externalities,
+                true,
+                "Metadata_metadata_versions",
+                &[],
+            )
+            .map_err(|_| {
+                CodegenError::Wasm("method \"Metadata_metadata_versions\" doesnt exist".to_owned())
+            })?;
+        let versions = <Vec<u32>>::decode(&mut &version[..])
+            .map_err(CodegenError::Decode)
+            .map(|x| {
+                x.into_iter()
+                    .filter(|version| SUPPORTED_METADATA_VERSIONS.contains(version))
+                    .collect::<Vec<u32>>()
+            })?;
+
+        if versions.is_empty() {
+            return Err(CodegenError::Other(
+                "No supported metadata versions were returned".to_owned(),
+            ));
+        }
+
+        Ok(versions)
+    }
+
+    fn metadata(&mut self) -> WasmMetadataResult<Metadata> {
```

Comment:
Nit: I'd call this something like `load_legacy_metadata` or something to be consistent with below (and make clear it's the "old" call :))

Reply:
renamed to `load_legacy_metadata`